### ISSUE TITLE
Beetle Adventure Racing, game slows down after time

### ISSUE THREAD

**Issue by orbea** [loganmc10/GLupeN64#56]
OS: `Slackware64-current`
`RetroArch-2016.09.11_ebeb081_master-x86_64-1_git`
`GLupeN64-2016.09.13_19b2c02_master-x86_64-1_git`

I noticed when playing Beetle Adventure Race the game will start lagging after time. For example if I start a novice circuit, the game will run full speed in the first race, but by time I get to the second race it will start having audio static which is soon followed by actual lag during the gameplay. This progressively increased in intensity until it became nearly unplayable in the third and final race. However just before I gave up the game froze for a few seconds and then resumed playing at full speed, but by time the third race was complete I noticed it was starting to have the audio static again.

**Comment by loganmc10** [comment]
Does toggling full screen do anything for you?

**Comment by orbea** [comment]
Yes and no. 

I was able to reproduce it the first try so I waited until the third race when it started lagging badly and toggled full screen. It resolved the lag immediately and I was able to continue playing until the third lap where it started again, but this it went from audio static to full system freeze keyboard and all within a matter of seconds.

I was able to ssh into my desktop from my laptop and found out that firefox was choking a cpu for some reason. So with htop I killed both RetroArch and firefox  which caused my ssh connection to immediately die forcing me to hard reset...

Here is my syslog from around the time of the system freeze, I was unable to check dmesg before I lost my ssh connection however.
```
Sep 14 15:44:10 Linux kernel: [875076.366325] BUG: unable to handle kernel NULL pointer dereference at 0000000000000030
Sep 14 15:44:10 Linux kernel: [875076.366387] IP: [<ffffffffc02e339b>] ttm_tt_swapout+0x4b/0x2b0 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.366427] PGD 0 
Sep 14 15:44:10 Linux kernel: [875076.366445] Oops: 0000 [#1] SMP 
Sep 14 15:44:10 Linux kernel: [875076.366469] Modules linked in: zram zsmalloc cpufreq_ondemand cfg80211 rfkill fuse snd_hda_codec_hdmi joydev mxm_wmi i2c_dev snd_hda_codec_realtek snd_hda_codec_generic evdev hid_microsoft snd_hda_intel kvm_amd snd_hda_codec kvm snd_hda_core nouveau(O) video ttm drm_kms_helper snd_hwdep snd_pcm snd_timer irqbypass drm crct10dif_pclmul crc32_pclmul serio_raw crc32c_intel agpgart k10temp ghash_clmulni_intel fam15h_power snd hwmon tpm_infineon wmi r8168(O) fb_sys_fops syscopyarea sysfillrect sysimgblt 8250_fintek i2c_algo_bit soundcore i2c_piix4 i2c_core button acpi_cpufreq tpm_tis tpm processor loop
Sep 14 15:44:10 Linux kernel: [875076.366980] CPU: 1 PID: 18073 Comm: kworker/u12:1 Tainted: G           O    4.6.7 #1
Sep 14 15:44:10 Linux kernel: [875076.367014] Hardware name: Gigabyte Technology Co., Ltd. To be filled by O.E.M./990FXA-UD5 R5, BIOS F3 04/01/2015
Sep 14 15:44:10 Linux kernel: [875076.367056] Workqueue: ttm_swap ttm_shrink_work [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367085] task: ffff880089444b00 ti: ffff88003d918000 task.ti: ffff88003d918000
Sep 14 15:44:10 Linux kernel: [875076.367115] RIP: 0010:[<ffffffffc02e339b>]  [<ffffffffc02e339b>] ttm_tt_swapout+0x4b/0x2b0 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367160] RSP: 0018:ffff88003d91bce8  EFLAGS: 00010206
Sep 14 15:44:10 Linux kernel: [875076.367184] RAX: 0000000000000000 RBX: ffff8801ec477400 RCX: 00000000006d5052
Sep 14 15:44:10 Linux kernel: [875076.367215] RDX: ffffffff00000001 RSI: ffff880220d39100 RDI: ffff880085abc780
Sep 14 15:44:10 Linux kernel: [875076.367246] RBP: ffff88003d91bd30 R08: ffff88003d91bcc0 R09: 0000000139b577d0
Sep 14 15:44:10 Linux kernel: [875076.367274] R10: 0000000000000001 R11: 000000000000000d R12: ffff880085abc780
Sep 14 15:44:10 Linux kernel: [875076.367303] R13: ffff8801ec477434 R14: ffff8800ce7c6590 R15: ffff8800ce7c6440
Sep 14 15:44:10 Linux kernel: [875076.367334] FS:  00007fa0b2d45e80(0000) GS:ffff88022ec40000(0000) knlGS:00000000f1da3b40
Sep 14 15:44:10 Linux kernel: [875076.367366] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Sep 14 15:44:10 Linux kernel: [875076.367392] CR2: 0000000000000030 CR3: 0000000001c06000 CR4: 00000000000406e0
Sep 14 15:44:10 Linux kernel: [875076.367420] Stack:
Sep 14 15:44:10 Linux kernel: [875076.367435]  ffff8801ec477400 ffff880220d39100 ffff88003d91bd10 ffffffffc02e4d89
Sep 14 15:44:10 Linux kernel: [875076.367494]  ffff8801ec477400 0000000000000000 ffff8801ec477434 ffff8800ce7c6590
Sep 14 15:44:10 Linux kernel: [875076.367549]  ffff8800ce7c6440 ffff88003d91bdc8 ffffffffc02e665b ffff880089444b80
Sep 14 15:44:10 Linux kernel: [875076.367601] Call Trace:
Sep 14 15:44:10 Linux kernel: [875076.367635]  [<ffffffffc02e4d89>] ? ttm_bo_unmap_virtual_locked+0x49/0x50 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367669]  [<ffffffffc02e665b>] ttm_bo_swapout+0x1fb/0x270 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367703]  [<ffffffff810bf774>] ? dequeue_task_fair+0x534/0x970
Sep 14 15:44:10 Linux kernel: [875076.367733]  [<ffffffff810c3851>] ? pick_next_task_fair+0x161/0x4b0
Sep 14 15:44:10 Linux kernel: [875076.367760]  [<ffffffffc02e27a4>] ttm_shrink+0xa4/0xd0 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367785]  [<ffffffffc02e27e9>] ttm_shrink_work+0x19/0x20 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.367811]  [<ffffffff810a4fc8>] process_one_work+0x148/0x4b0
Sep 14 15:44:10 Linux kernel: [875076.367837]  [<ffffffff810a5378>] worker_thread+0x48/0x4d0
Sep 14 15:44:10 Linux kernel: [875076.367863]  [<ffffffff810a5330>] ? process_one_work+0x4b0/0x4b0
Sep 14 15:44:10 Linux kernel: [875076.367890]  [<ffffffff810aa799>] kthread+0xc9/0xe0
Sep 14 15:44:10 Linux kernel: [875076.367913]  [<ffffffff817bff02>] ret_from_fork+0x22/0x40
Sep 14 15:44:10 Linux kernel: [875076.367936]  [<ffffffff810aa6d0>] ? kthread_worker_fn+0x170/0x170
Sep 14 15:44:10 Linux kernel: [875076.367960] Code: e5 41 57 41 56 41 55 41 54 49 89 fc 53 48 83 ec 20 48 85 f6 48 89 75 c0 0f 84 db 01 00 00 48 8b 45 c0 48 83 7f 28 00 48 8b 40 20 <4c> 8b 78 30 0f 84 31 02 00 00 48 8b 45 c0 48 89 45 b8 65 48 8b 
Sep 14 15:44:10 Linux kernel: [875076.368486] RIP  [<ffffffffc02e339b>] ttm_tt_swapout+0x4b/0x2b0 [ttm]
Sep 14 15:44:10 Linux kernel: [875076.368517]  RSP <ffff88003d91bce8>
Sep 14 15:44:10 Linux kernel: [875076.368532] CR2: 0000000000000030
Sep 14 15:44:10 Linux kernel: [875076.422937] ---[ end trace 887e861ddd6fca31 ]---
Sep 14 15:44:10 Linux kernel: [875076.423038] BUG: unable to handle kernel paging request at ffffffffffffffd8
Sep 14 15:44:10 Linux kernel: [875076.423097] IP: [<ffffffff810aae00>] kthread_data+0x10/0x20
Sep 14 15:44:10 Linux kernel: [875076.423129] PGD 1c07067 PUD 1c09067 PMD 0 
Sep 14 15:44:10 Linux kernel: [875076.423160] Oops: 0000 [#2] SMP 
Sep 14 15:44:10 Linux kernel: [875076.423183] Modules linked in: zram zsmalloc cpufreq_ondemand cfg80211 rfkill fuse snd_hda_codec_hdmi joydev mxm_wmi i2c_dev snd_hda_codec_realtek snd_hda_codec_generic evdev hid_microsoft snd_hda_intel kvm_amd snd_hda_codec kvm snd_hda_core nouveau(O) video ttm drm_kms_helper snd_hwdep snd_pcm snd_timer irqbypass drm crct10dif_pclmul crc32_pclmul serio_raw crc32c_intel agpgart k10temp ghash_clmulni_intel fam15h_power snd hwmon tpm_infineon wmi r8168(O) fb_sys_fops syscopyarea sysfillrect sysimgblt 8250_fintek i2c_algo_bit soundcore i2c_piix4 i2c_core button acpi_cpufreq tpm_tis tpm processor loop
Sep 14 15:44:10 Linux kernel: [875076.423653] CPU: 1 PID: 18073 Comm: kworker/u12:1 Tainted: G      D    O    4.6.7 #1
Sep 14 15:44:10 Linux kernel: [875076.423687] Hardware name: Gigabyte Technology Co., Ltd. To be filled by O.E.M./990FXA-UD5 R5, BIOS F3 04/01/2015
Sep 14 15:44:10 Linux kernel: [875076.423742] task: ffff880089444b00 ti: ffff88003d918000 task.ti: ffff88003d918000
Sep 14 15:44:10 Linux kernel: [875076.423776] RIP: 0010:[<ffffffff810aae00>]  [<ffffffff810aae00>] kthread_data+0x10/0x20
Sep 14 15:44:10 Linux kernel: [875076.423817] RSP: 0018:ffff88003d91b9e8  EFLAGS: 00010002
Sep 14 15:44:10 Linux kernel: [875076.439137] RAX: 0000000000000000 RBX: 0000000000014dc0 RCX: ffffffff81e8e040
Sep 14 15:44:10 Linux kernel: [875076.439168] RDX: 0000000000000001 RSI: ffff880089444b80 RDI: ffff880089444b00
Sep 14 15:44:10 Linux kernel: [875076.439193] RBP: ffff88003d91b9e8 R08: 0000000000000000 R09: 0000000000000000
Sep 14 15:44:10 Linux kernel: [875076.439216] R10: 0000000000007000 R11: ffff880089444b80 R12: ffff88022ec54dc0
Sep 14 15:44:10 Linux kernel: [875076.439241] R13: 0000000000000000 R14: ffff880089444fe0 R15: ffff880089444b00
Sep 14 15:44:10 Linux kernel: [875076.439265] FS:  00007fa0b2d45e80(0000) GS:ffff88022ec40000(0000) knlGS:00000000f1da3b40
Sep 14 15:44:10 Linux kernel: [875076.439291] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
Sep 14 15:44:10 Linux kernel: [875076.439310] CR2: 0000000000000028 CR3: 0000000001c06000 CR4: 00000000000406e0
Sep 14 15:44:10 Linux kernel: [875076.439333] Stack:
Sep 14 15:44:10 Linux kernel: [875076.439343]  ffff88003d91b9f8 ffffffff810a5c1e ffff88003d91ba50 ffffffff817bbdf2
Sep 14 15:44:10 Linux kernel: [875076.457108]  0000000000000000 ffff880089444b00 ffff8800ac1c8b08 ffff88003d91ba50
Sep 14 15:45:10 Linux kernel: [875076.489498]  ffff88003d91c000 ffff88003d91baa0 ffff88003d91b698 ffff880089444b00
Sep 14 15:45:10 Linux kernel: [875076.506836] Call Trace:
Sep 14 15:45:10 Linux kernel: [875076.539610]  [<ffffffff810a5c1e>] wq_worker_sleeping+0xe/0x90
Sep 14 15:45:10 Linux kernel: [875076.556896]  [<ffffffff817bbdf2>] __schedule+0x552/0xa30
Sep 14 15:45:10 Linux kernel: [875076.589660]  [<ffffffff817bc305>] schedule+0x35/0x80
Sep 14 15:45:10 Linux kernel: [875076.606869]  [<ffffffff8109088b>] do_exit+0x77b/0xb10
Sep 14 15:45:10 Linux kernel: [875076.608323]  [<ffffffff8101f56c>] oops_end+0x6c/0x90
Sep 14 15:45:10 Linux kernel: [875076.610101]  [<ffffffff81047e6e>] no_context+0x11e/0x370
Sep 14 15:45:10 Linux kernel: [875076.611621]  [<ffffffff8104814a>] __bad_area_nosemaphore+0x8a/0x1f0
Sep 14 15:45:10 Linux kernel: [875076.613060]  [<ffffffff810482c4>] bad_area_nosemaphore+0x14/0x20
Sep 14 15:45:10 Linux kernel: [875076.614799]  [<ffffffff810485ce>] __do_page_fault+0x8e/0x3f0
Sep 14 15:45:10 Linux kernel: [875076.616325]  [<ffffffff8119202c>] ? tlb_finish_mmu+0x1c/0x50
Sep 14 15:45:10 Linux kernel: [875076.617679]  [<ffffffff8104896c>] do_page_fault+0xc/0x10
Sep 14 15:45:10 Linux kernel: [875076.619379]  [<ffffffff817c1822>] page_fault+0x22/0x30
Sep 14 15:45:10 Linux kernel: [875076.620886]  [<ffffffffc02e339b>] ? ttm_tt_swapout+0x4b/0x2b0 [ttm]
Sep 14 15:45:10 Linux kernel: [875076.622259]  [<ffffffffc02e4d89>] ? ttm_bo_unmap_virtual_locked+0x49/0x50 [ttm]
Sep 14 15:45:10 Linux kernel: [875076.623949]  [<ffffffffc02e665b>] ttm_bo_swapout+0x1fb/0x270 [ttm]
Sep 14 15:45:10 Linux kernel: [875076.625461]  [<ffffffff810bf774>] ? dequeue_task_fair+0x534/0x970
Sep 14 15:45:10 Linux kernel: [875076.627027]  [<ffffffff810c3851>] ? pick_next_task_fair+0x161/0x4b0
Sep 14 15:45:10 Linux kernel: [875076.628583]  [<ffffffffc02e27a4>] ttm_shrink+0xa4/0xd0 [ttm]
Sep 14 15:45:10 Linux kernel: [875076.630200]  [<ffffffffc02e27e9>] ttm_shrink_work+0x19/0x20 [ttm]
Sep 14 15:45:10 Linux kernel: [875076.631693]  [<ffffffff810a4fc8>] process_one_work+0x148/0x4b0
Sep 14 15:45:10 Linux kernel: [875076.633246]  [<ffffffff810a5378>] worker_thread+0x48/0x4d0
Sep 14 15:45:10 Linux kernel: [875076.634642]  [<ffffffff810a5330>] ? process_one_work+0x4b0/0x4b0
Sep 14 15:45:10 Linux kernel: [875076.636093]  [<ffffffff810aa799>] kthread+0xc9/0xe0
Sep 14 15:45:10 Linux kernel: [875076.637394]  [<ffffffff817bff02>] ret_from_fork+0x22/0x40
Sep 14 15:45:10 Linux kernel: [875076.638719]  [<ffffffff810aa6d0>] ? kthread_worker_fn+0x170/0x170
Sep 14 15:45:10 Linux kernel: [875076.640039] Code: ff ff ff 75 d8 48 8d 7c 24 18 e8 cc 26 71 00 e9 5f ff ff ff 0f 1f 80 00 00 00 00 0f 1f 44 00 00 48 8b 87 80 04 00 00 55 48 89 e5 <48> 8b 40 d8 5d c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 
Sep 14 15:45:10 Linux kernel: [875076.643685] RIP  [<ffffffff810aae00>] kthread_data+0x10/0x20
Sep 14 15:45:10 Linux kernel: [875076.645305]  RSP <ffff88003d91b9e8>
Sep 14 15:45:10 Linux kernel: [875076.646816] CR2: ffffffffffffffd8
Sep 14 15:45:10 Linux kernel: [875076.648623] ---[ end trace 887e861ddd6fca32 ]---
Sep 14 15:45:10 Linux kernel: [875076.648624] Fixing recursive fault but reboot is needed!
Sep 14 15:45:10 Linux kernel: [875136.653324] INFO: rcu_sched detected stalls on CPUs/tasks:
Sep 14 15:45:10 Linux kernel: [875136.653329] ^I1-...: (2 GPs behind) idle=bc9/140000000000000/0 softirq=63298776/63298777 fqs=19987 
Sep 14 15:45:10 Linux kernel: [875136.653333] ^I4-...: (1 GPs behind) idle=3f3/140000000000001/0 softirq=134498667/134498667 fqs=19987 
Sep 14 15:45:10 Linux kernel: [875136.653335] ^I5-...: (1 GPs behind) idle=749/1/0 softirq=66826542/66826543 fqs=19987 
Sep 14 15:45:10 Linux kernel: [875136.653336] ^I(detected by 0, t=60002 jiffies, g=129957635, c=129957634, q=8409)
Sep 14 15:45:10 Linux kernel: [875136.653351]  0000000000000010 0000000000010002 ffff88003d91b9e8 0000000000000018
Sep 14 15:45:10 Linux kernel: [875136.653354]  ffff88003d91b9f8 ffff88003d91b9f8 ffffffff810a5c1e ffff88003d91ba50
Sep 14 15:45:10 Linux kernel: [875136.653356]  ffffffff817bbdf2 0000000000000000 ffff880089444b00 ffff8800ac1c8b08
Sep 14 15:45:10 Linux kernel: [875136.653358] Call Trace:
Sep 14 15:45:10 Linux kernel: [875136.653364]  [<ffffffff810a5c1e>] ? wq_worker_sleeping+0xe/0x90
Sep 14 15:45:10 Linux kernel: [875136.653368]  [<ffffffff817bbdf2>] ? __schedule+0x552/0xa30
Sep 14 15:45:10 Linux kernel: [875136.653370]  [<ffffffff817bc305>] ? schedule+0x35/0x80
Sep 14 15:45:10 Linux kernel: [875136.653372]  [<ffffffff8109088b>] ? do_exit+0x77b/0xb10
Sep 14 15:45:10 Linux kernel: [875136.653375]  [<ffffffff8101f56c>] ? oops_end+0x6c/0x90
Sep 14 15:45:10 Linux kernel: [875136.653377]  [<ffffffff81047e6e>] ? no_context+0x11e/0x370
Sep 14 15:45:10 Linux kernel: [875136.653379]  [<ffffffff8104814a>] ? __bad_area_nosemaphore+0x8a/0x1f0
Sep 14 15:45:10 Linux kernel: [875136.653381]  [<ffffffff810482c4>] ? bad_area_nosemaphore+0x14/0x20
Sep 14 15:45:10 Linux kernel: [875136.653383]  [<ffffffff810485ce>] ? __do_page_fault+0x8e/0x3f0
Sep 14 15:45:10 Linux kernel: [875136.653386]  [<ffffffff8119202c>] ? tlb_finish_mmu+0x1c/0x50
Sep 14 15:45:10 Linux kernel: [875136.653388]  [<ffffffff8104896c>] ? do_page_fault+0xc/0x10
Sep 14 15:45:10 Linux kernel: [875136.653390]  [<ffffffff817c1822>] ? page_fault+0x22/0x30
Sep 14 15:45:10 Linux kernel: [875136.653396]  [<ffffffffc02e339b>] ? ttm_tt_swapout+0x4b/0x2b0 [ttm]
Sep 14 15:45:10 Linux kernel: [875136.653399]  [<ffffffffc02e4d89>] ? ttm_bo_unmap_virtual_locked+0x49/0x50 [ttm]
Sep 14 15:45:10 Linux kernel: [875136.653402]  [<ffffffffc02e665b>] ? ttm_bo_swapout+0x1fb/0x270 [ttm]
Sep 14 15:45:10 Linux kernel: [875136.653405]  [<ffffffff810bf774>] ? dequeue_task_fair+0x534/0x970
Sep 14 15:45:10 Linux kernel: [875136.653407]  [<ffffffff810c3851>] ? pick_next_task_fair+0x161/0x4b0
Sep 14 15:45:10 Linux kernel: [875136.653410]  [<ffffffffc02e27a4>] ? ttm_shrink+0xa4/0xd0 [ttm]
Sep 14 15:45:10 Linux kernel: [875136.653412]  [<ffffffffc02e27e9>] ? ttm_shrink_work+0x19/0x20 [ttm]
Sep 14 15:45:10 Linux kernel: [875136.653414]  [<ffffffff810a4fc8>] ? process_one_work+0x148/0x4b0
Sep 14 15:45:10 Linux kernel: [875136.653416]  [<ffffffff810a5378>] ? worker_thread+0x48/0x4d0
Sep 14 15:45:10 Linux kernel: [875136.653417]  [<ffffffff810a5330>] ? process_one_work+0x4b0/0x4b0
Sep 14 15:45:10 Linux kernel: [875136.653419]  [<ffffffff810aa799>] ? kthread+0xc9/0xe0
Sep 14 15:45:10 Linux kernel: [875136.653422]  [<ffffffff817bff02>] ? ret_from_fork+0x22/0x40
Sep 14 15:45:10 Linux kernel: [875136.653424]  [<ffffffff810aa6d0>] ? kthread_worker_fn+0x170/0x170
Sep 14 15:45:10 Linux kernel: [875136.653429]  ffffffff817bbbbf ffffffff81c0b500 0000000000000001 ffff8801583b3cd8
Sep 14 15:45:10 Linux kernel: [875136.653431]  ffffffff811b13a9 ffff8801583b4000 ffff8800cb886400 0000000000000000
Sep 14 15:45:10 Linux kernel: [875136.653433]  ffff8801583b3d60 ffff88001353abe0 ffff88001353abe0 ffffffff817bc305
Sep 14 15:45:10 Linux kernel: [875136.653436] Call Trace:
Sep 14 15:45:10 Linux kernel: [875136.653438]  [<ffffffff817bbbbf>] ? __schedule+0x31f/0xa30
Sep 14 15:45:10 Linux kernel: [875136.653440]  [<ffffffff811b13a9>] ? __slab_free+0x99/0x2c0
Sep 14 15:45:10 Linux kernel: [875136.653442]  [<ffffffff817bc305>] ? schedule+0x35/0x80
Sep 14 15:45:10 Linux kernel: [875136.653444]  [<ffffffff811b179d>] ? kmem_cache_free+0x1cd/0x1e0
Sep 14 15:45:10 Linux kernel: [875136.653446]  [<ffffffff81098353>] ? __sigqueue_free.part.15+0x33/0x40
Sep 14 15:45:10 Linux kernel: [875136.653448]  [<ffffffff810bbf9b>] ? update_curr+0xcb/0x140
Sep 14 15:45:10 Linux kernel: [875136.653451]  [<ffffffff810feb0f>] ? get_futex_key+0x27f/0x330
Sep 14 15:45:10 Linux kernel: [875136.653454]  [<ffffffff810b5ea0>] ? task_sched_runtime+0xb0/0x120
Sep 14 15:45:10 Linux kernel: [875136.653455]  [<ffffffff810b9f39>] ? cputime_adjust+0x89/0x140
Sep 14 15:45:10 Linux kernel: [875136.653457]  [<ffffffff810ba4b8>] ? thread_group_cputime_adjusted+0x48/0x60
Sep 14 15:45:10 Linux kernel: [875136.653460]  [<ffffffff810896e2>] ? mmput+0x12/0xf0
Sep 14 15:45:10 Linux kernel: [875136.653462]  [<ffffffff8109d8e6>] ? k_getrusage+0x3b6/0x3c0
Sep 14 15:45:10 Linux kernel: [875136.653464]  [<ffffffff810a0690>] ? getrusage+0x30/0x50
Sep 14 15:45:10 Linux kernel: [875136.653466]  [<ffffffff810a06e1>] ? SyS_getrusage+0x31/0x40
Sep 14 15:45:10 Linux kernel: [875136.653468]  [<ffffffff817bfcdf>] ? entry_SYSCALL_64_fastpath+0x17/0x93
Sep 14 15:45:10 Linux kernel: [875136.653473]  0000000000000246 ffff880225a1fe88 0000000000000018 ffffffff8168233c
Sep 14 15:45:10 Linux kernel: [875136.653475]  0000000000000000 00031bde13806268 ffff880225a20000 ffff88022407c600
Sep 14 15:45:10 Linux kernel: [875136.653477]  ffffffffc000e880 ffff880225a1c000 ffff880225a1c000 ffff880225a1fed0
Sep 14 15:45:10 Linux kernel: [875136.653479] Call Trace:
Sep 14 15:45:10 Linux kernel: [875136.653482]  [<ffffffff8168233c>] ? cpuidle_enter_state+0x12c/0x2e0
Sep 14 15:45:10 Linux kernel: [875136.653484]  [<ffffffff81682527>] ? cpuidle_enter+0x17/0x20
Sep 14 15:45:10 Linux kernel: [875136.653486]  [<ffffffff810ca32a>] ? call_cpuidle+0x2a/0x40
Sep 14 15:45:10 Linux kernel: [875136.653488]  [<ffffffff810ca6e5>] ? cpu_startup_entry+0x285/0x330
Sep 14 15:45:10 Linux kernel: [875136.653490]  [<ffffffff8103a756>] ? start_secondary+0x136/0x140
```

**Comment by orbea** [comment]
This appears to be a kernel issue, specifically concerning ttm / nouveau.

**Comment by loganmc10** [comment]
So would you classify this as an issue that happens after playing for a while? Or it happens at a certain stage in the game (could it be reproduced from a save state)?

GLideN64 keeps a texture cache, as the game progresses and it comes across new textures, it stores them in the cache until the cache hits 500MB in size, or 128,000 textures. Some games (GoldenEye and Beetle Adventure Racing come to mind) seem to generate never-ending textures. I think this leads to slowdowns overtime. When you toggle full screen, it resets the OpenGL context, basically resetting everything, including the texture cache.

Basically what I'm saying is that I've noticed this in GoldenEye with mupen64plus+GLideN64, but toggling full screen should reset things for you. I'm not sure why your machine crapped all over itself though.

If it is reproducible thought (happens at a certain stage no longer how long you've been playing) I'll take a look if you have a save state.

**Comment by orbea** [comment]
Trying to see how it reacts to save states was the next step, but I suspect it to be related to how long I have been playing rather than a specific point in the stage. For example if I play a single race this does not happen, but if I start playing more than one race in a row it will occur. Additionally some stages may have more textures being rendered than others and seem to trigger it more often.

My understanding there is a problem with how Beetle Adventure Racing + GLideN64 is reacting with the kernel ttm + nouveau code. How much should be fixed in either the kernel or in GLideN64 is hard for me to tell, I think the best way to find that out would if someone far more familiar with GLideN64 than myself tried to discuss this with nouveau developers. This can be done in #nouveau @ freenode.

Another consideration is that I should find a network card that works with a driver in the kernel so I can check if this has been fixed in Linux 4.7 or 4.8.

**Comment by loganmc10** [comment]
There is the potential that there is something wrong with the ```TextureCache::_clear()``` function in GLideN64, I'll play around with it a bit, it probably doesn't get tested much since it only gets triggered if the cache hits 500MB or 128000 textures

**Comment by loganmc10** [comment]
It all seems to be working probably, anyway the slowdown is basically a GLideN64 thing that you'll notice with a few games, but toggling full screen should clear it up for a while.

GLideN64 has a number of performance issues, if you look through the GitHub issues you see tons of issues related to performance.

It seems like it's just related to the size of the texture cache, in the past there have been issues with texture corruption because these games generate so many textures, the corruption has mostly been sorted out, but the cache still gets big:
https://github.com/gonetz/GLideN64/issues/744
https://github.com/gonetz/GLideN64/issues/459

Also a comment from someone who seems to have run into the same issue as you:
https://github.com/gonetz/GLideN64/pull/745#issuecomment-145862678

**Comment by orbea** [comment]
From talking to the nouveau devs, the system crash was an error with swapping something out of or back into vram.

I checked and this is not reproducible from a save state.

**Comment by orbea** [comment]
So this is related to the size of the texture cache. I was able to mitigate the issue by reducing the texture cache max size by 75%.
```
void TextureCache::_checkCacheSize()
{
#ifdef VC
        const size_t maxCacheSize = 3750;
#else
        const size_t maxCacheSize = 32000;
#endif
```
This entirely avoided the slow down, but as a result the game would occasionally freeze (Hiccup) for a moment, presumably when clearing the textures cache. Otherwise I was able to complete all 3 races in the novice tournament with no game issues or system crashes.

**Comment by loganmc10** [comment]
I figured as much, the VC number shouldn't be changed, that was determined based on testing (https://github.com/gonetz/GLideN64/pull/1017) for the Raspberry Pi.

I'm not sure where gonetz came up with the 128,000 number. I assume a number like 32,000 would be a lot safer for most devices, and most games will never hit that anyway (only games like Beetle or GoldenEye that seem to have this problem), but it's still better than freezing and crazy slowdowns.

**Comment by loganmc10** [comment]
Also, you'll see in that pull request (https://github.com/gonetz/GLideN64/pull/1017/files) I wrote it to delete the oldest texture when it hit the limit. For example, once it hit 15000 (this is for the Raspberry Pi) it would delete the texture that hadn't been used in the longest time in order to make room for the new texture, keeping the number at 15000.

For some reason after a few weeks gonetz re-wrote it to just wipe the whole cache if it hit the limit (and that's when he added the 128000 limit for regular devices). This is what causes the hiccup, if it was reimplemented to just delete the oldest texture and replace it with a new one, keeping the cache at it's limit, that would get rid of the hiccup and stop the cache from getting too big.

**Comment by loganmc10** [comment]
Here is where it was re-written:

https://github.com/gonetz/GLideN64/commit/803eb1d3814e469b811f3b60b5df35d8197b71f6

**Comment by orbea** [comment]
It seems that nouveau may be converting each texture into a 128 KB "large" page, so at 128,000 textures that would be 16 GB when my vram is 3 GB... It was suggested to try reducing 128000 to 16384 (2 GB), this both increased the frequency and decreased the severity of the hiccups and may be a sane number. I will try your original change in a bit to see if it helps the hiccups.
```
void TextureCache::_checkCacheSize()
{
#ifdef VC
        const size_t maxCacheSize = 15000;
#else
        const size_t maxCacheSize = 16384;
#endif 
```

**Comment by loganmc10** [comment]
Ya just replace
```
_clear();
return;
```

with
```
CachedTexture& piTex = m_textures.back();
m_cachedBytes -= piTex.textureBytes;
glDeleteTextures(1, &piTex.glName);
m_lruTextureLocations.erase(piTex.crc);
m_textures.pop_back();
```

**Comment by loganmc10** [comment]
Before that commit gliden64 had no limit, just a MB size limit set to 500MB by default (but like you mention, that is 500MB in GLideN64's eyes, the driver might actually be storing more).

That commit introduced a "number" limit of 128000 textures (15000 on the Raspberry Pi)

**Comment by orbea** [comment]
I've tested this again with that changes, there are no slowdowns, hiccups or crashes. I would consider this a complete solution. Thanks!
```
#ifdef VC
        const size_t maxCacheSize = 15000;
#else                                     
        const size_t maxCacheSize = 16384;
#endif
        // Clear cache if its size is too large.
        if (m_textures.size() >= maxCacheSize) {
                CachedTexture& piTex = m_textures.back();
                m_cachedBytes -= piTex.textureBytes;
                glDeleteTextures(1, &piTex.glName);
                m_lruTextureLocations.erase(piTex.crc);
                m_textures.pop_back();
        }
```
I would suggest closing this issue in the commit that merges this fix from GLideN64, but if you want to close this sooner go ahead.

**Comment by loganmc10** [comment]
I'll leave this open until it is fixed in GLideN64, are you going to submit this? The only thing I would change is remove the ```// Clear cache if its size is too large.``` since it's not actually clearing the cache anymore.

**Comment by orbea** [comment]
I submitted a PR.

I changed `piTex` to `clsTex` which as was informed by a friend refers to a dos command that would clear the screen. It seemed at least somewhat fitting.
